new logger - initial version

mgrybyk opened the pull request:
## Proposed changes

Removed usage of **log-update**

![image](https://user-images.githubusercontent.com/25589559/57389373-4b23bd80-71ba-11e9-9312-33b701dfa83b.png)

**Videos**
- [single instance](https://drive.google.com/file/d/1ahxkqm8GDM5uObuN6UMjeRYrqK1mjtxr/view?usp=sharing)
- [2 instances](https://drive.google.com/file/d/1456DUehcSRuxl42uGITCCilICNPYT8EW/view?usp=sharing)

should fix these ones: #3826 #3869

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

**Added tests**

Styling suggestions are really welcomed!

### Reviewers: @webdriverio/technical-committee

mgrybyk commented:
added videos

abjerstedt commented:
So we are still not printing the reporters until the very end of the run?

mgrybyk commented:
Not, but I can change that in one condition or make it configurable. 
As you like.
Currently reporters are printed straight away in watch mode only

caoxu2000 commented:
why the 2nd test '05.ts' results got printed twice?

mgrybyk commented:
Because it was retried and failed both times

CrispusDH commented on packages/wdio-cli/src/interface.js:
```diff
@@ -55,19 +49,58 @@ export default class WDIOCLInterface extends EventEmitter {
+
+    onJobComplete(cid, job, retries, message) {
+        const details = []
+        if (job) {
```
Do you suggest that `job` could be not exist?

mgrybyk replied:
Yeah, there is such test on launcher.test.js, not sure how to reproduce this in real life.
In one of the cases job doesn't exist

mgrybyk replied:
@CrispusDH this test is failing https://github.com/webdriverio/webdriverio/blob/8de7f1a3b12d97282ed4ee2e25e4c3c9a8940ac1/packages/wdio-cli/tests/launcher.test.js#L105

job is undefined in this test both for old and current interface implementations.
It was added by @bennieswart 3 months ago, not sure how to get such situation in real life.

CrispusDH commented on packages/wdio-cli/src/interface.js:
```diff
@@ -164,204 +180,37 @@ export default class WDIOCLInterface extends EventEmitter {
-        this.printSummary()
-        this.updateClock()
+        const isFinished = this.jobs.size === 0
+        const shutdownMessage = !isFinished ? 'Ending WebDriver sessions gracefully ...\n' + '(press ctrl+c again to hard kill the runner)' : 'Ended WebDriver sessions gracefully after a SIGINT signal was received!'
```
It might be better use `isRunning` instead of `!isFinished`

mgrybyk replied:
I'll check, I have copied most of the logic, like these two lines.

mgrybyk replied:
changed to isRunning

CrispusDH commented on packages/wdio-cli/src/interface.js:
```diff
@@ -55,19 +48,58 @@ export default class WDIOCLInterface extends EventEmitter {
+        if (job) {
+            const filename = Array.isArray(job.specs) ? job.specs.join(', ').replace(process.cwd(), '') : ''
+            const retryCount = retries > 0 ? `(${retries} retries)` : ''
+            details.push('in', getRunnerName(job.caps), filename ? `- ${filename}` : '', retryCount)
```
It might be cleaner to put as argument just `job` variable and change `getRunnerName` to:
```js
export function getRunnerName ({caps = {}}) {
   ...
}
```

mgrybyk replied:
Can't get what do you mean here. I can't modify getRunnerName function that way. It accepts capabilities object, not an object containing capabilities.
```{
 browserName: 'Chrome'
}```
not
```{
  capabilities: {
     browserName: 'Chrome'
  }
}
```

CrispusDH replied:
but you insert argument like `job.caps`. So, if you will insert `job` object with `caps` property it will work.

mgrybyk replied:
I think we are talking about different things here, getRunnerName doesn't accept job, it accepts capabilities

CrispusDH commented on packages/wdio-cli/src/interface.js:
```diff
@@ -55,19 +48,58 @@ export default class WDIOCLInterface extends EventEmitter {
+    onJobComplete(cid, job, retries, message) {
+        const details = []
+        if (job) {
+            const filename = Array.isArray(job.specs) ? job.specs.join(', ').replace(process.cwd(), '') : ''
```
I think it makes sense to create a function for getting `filename` and then remove the inline variable completely. In this case we will get cleaner main function and remove condition in line 87.
```js
function filename({specs}) {
   const result = Array.isArray(specs) ? specs.join(', ').replace(process.cwd(), '') : '';
   return result ? `- ${result}` : ''
}
```
after it line 87 would look like:
`details.push('in', getRunnerName(job.caps), filename(job), retryCount)`

mgrybyk replied:
@CrispusDH please check again

erwinheitzman commented:
> Because it was retried and failed both times.
> 
> Do we need to change this? I'm not sure to be honest because we have really executed that file twice

We used to only show one result though if we want to change this, now would be a good time.

erwinheitzman commented:
Why are some of the tests skipped? If they are no longer relevant, remove them instead.

mgrybyk commented:
@erwinheitzman I haven't started working on the tests yet, was just reviewing existing. Want to implement other suggestions first

mgrybyk commented:
@erwinheitzman result was printed in case of retry before as well. 

![image](https://user-images.githubusercontent.com/25589559/57436004-1dcd2300-723f-11e9-8b0a-8b2c58a0771d.png)

CrispusDH commented:
Would it make a sense to have such algorithm:
1. Run all tests
2. Show list of passed and failed
3. Retry failed tests
4. Show list of passed and failed
....
In this case would be easier to get what tests were retried, what tests were failed twice, etc. Currently, it would be pretty hard to understand. For example you will get at the end such message:
```js
125 passed, 13 retries, 7 failed
```
So, after retry 6 of 13 first failed tests were passed passed and 7 were failed again. But what exactly? How to find it?

bennieswart commented:
I've been following this with half an ear since I got mentioned so I'm not up to date with the whole context, but I see there is some uncertainty about the retries. It might be of value to review the code here https://github.com/webdriverio/webdriverio/pull/3381, and specifically have a look at the fields added to `onRunnerStart` (`retry`) and `onRunnerEnd` (`retries`) here https://github.com/webdriverio/webdriverio/pull/3381/files?file-filters%5B%5D=.js&file-filters%5B%5D=.md#diff-f9fbbc3907f0d68d7bf074ad66870d03 .

If a runner starts with `retry > 0` then it has previously run and failed. If a runner fails with `retries > 0`, then it will be retried again.

I hope this helps.

bennieswart commented:
In reponse to @CrispusDH's suggestion, I'd prefer if tests were rescheduled asap upon failure, reason being that we take great care regarding the order of execution of our tests to run longest tests first, so as not to wait for a single long-running test at the end after all other tests have completed. When such a test fails the first time, it should immediately be rescheduled for the same reason.

mgrybyk commented:
Anyway runner logic is not a part of logger, if you'd like to change flow you can submit another ticket for that

mgrybyk commented:
What do you think regarding adding timestamp to workers output?

![image](https://user-images.githubusercontent.com/25589559/57514580-18400d80-7311-11e9-8b2d-c0aa96c33929.png)

mgrybyk commented:
PR updated, removed WIP label, added tests.

BorisOsipov commented on packages/wdio-cli/src/interface.js:
```diff
@@ -1,31 +1,22 @@
-        this.stdoutBuffer = []
-        this.stderr = stderr
-        this.stderrBuffer = []
+        this.specFileRetries = config.specFileRetries || 0
```
Yes in runner configuration docs

christian-bromann commented:
@mgrybyk will test this today

mgrybyk commented:
Tested logging on Jenkins and it works fine! In pipelines colored logs are shown in realtime without duplications. Same for regular jobs, but logs are not colored by default, added docs on how to use `FORCE_COLOR` in case there is some plugin like AnsiColors installed.

Also fixed issues with watch mode.

mgrybyk commented:
no changes made, just synced with master

christian-bromann commented:
@mgrybyk can you please rebase on top of master?

@webdriverio/project-committers can I get another 👍 so we can merge this?

mgrybyk commented:
rebasing in 10 minutes

mgrybyk commented:
@christian-bromann rebased, can be merged!
